Use Implicit User Principal Name If Explicit Does Not Exist

NoMoreFood opened the pull request:
- Modified user principal name lookup to default to the implicit form (SamAccountName@DnsDomainName) if no explicit user principal name attribute is found on the account.

NoMoreFood committed:
Use Implicit User Principal Name If Explicit Does Not Exist

- Modified user principal name lookup to default to the implicit form (SamAccountName@DnsDomainName) if no explicit user principal name attribute is found on the account.

manojampalam commented on contrib/win32/win32compat/win32_usertoken_utils.c:
```diff
@@ -715,4 +706,37 @@ get_custom_lsa_package()
+		wcscat_s(user_principal_name, MAX_UPN_LEN + 1, L"@");
+		wcsncat_s(user_principal_name, MAX_UPN_LEN + 1, domain_upn, wcschr(domain_upn, L'/') - domain_upn);
+		debug3("%s: Successfully discovered implicit principal name: '%ls'=>'%ls'",
+			__FUNCTION__, sam_account_name, user_principal_name);
```
help me understand. How would the NameCanonical name be different from NameSamCompatible name ? As per msdn NameCanonical can have multiple '/'s.

NoMoreFood replied:
We're only using the fully qualified domain name part of NameCanonical (i.e. **domain.local**/other/path/stuff).

manojampalam replied:
Thanks. Missed it earlier.

NoMoreFood replied:
Added some additional explanation to the code as well.

manojampalam commented on contrib/win32/win32compat/win32_usertoken_utils.c:
```diff
@@ -715,4 +706,37 @@ get_custom_lsa_package()
+	lookup_error = GetLastError();
+	domain_upn_len = ARRAYSIZE(domain_upn);
+	if (pTranslateNameW(sam_account_name, NameSamCompatible, NameCanonical, domain_upn, &domain_upn_len) != 0) {
+		wcscpy_s(user_principal_name, MAX_UPN_LEN + 1, wcschr(sam_account_name, L'\\') + 1);
```
Initialize user_principal_name to "\0".

NoMoreFood replied:
Not sure the intent here.  String copy functions do not need an initialized string since they start copying immediately into the passed pointer without looking at what data might be there.

manojampalam commented on contrib/win32/win32compat/win32_usertoken_utils.c:
```diff
@@ -715,4 +706,37 @@ get_custom_lsa_package()
 	return s_lsa_auth_pkg;
 }
 
+int lookup_principal_name(const wchar_t * sam_account_name, wchar_t * user_principal_name)
```
recommend making this a void func and adding a comment that it provides best effort upn resolution

manojampalam commented on contrib/win32/win32compat/win32_usertoken_utils.c:
```diff
@@ -715,4 +706,37 @@ get_custom_lsa_package()
+	/* report error and copy the sam account name into the buffer */
+	error("%s: User principal name lookup failed for user '%ls' (explicit: %d, implicit: %d)",
+		__FUNCTION__, sam_account_name, lookup_error, GetLastError());
+	wcscpy_s(user_principal_name, MAX_UPN_LEN + 1, sam_account_name);
```
remove this.. it will be clean to assume that nothing gets set on the failure path

NoMoreFood replied:
Done

manojampalam commented on contrib/win32/win32compat/win32_usertoken_utils.c:
```diff
@@ -133,19 +133,10 @@ generate_s4u_user_token(wchar_t* user_cpn, int impersonation) {
-		else
-			debug3("%s: Successfully discovered principal name: '%ls'=>'%ls'", __FUNCTION__, user_cpn, domain_upn);
-
+		lookup_principal_name(user_cpn, domain_upn);
```
handle failure (copy user_cpn to domain_upn on failure)

NoMoreFood replied:
Done

manojampalam commented on contrib/win32/win32compat/win32_usertoken_utils.c:
```diff
@@ -715,4 +706,37 @@ get_custom_lsa_package()
 	return s_lsa_auth_pkg;
 }
 
+int lookup_principal_name(const wchar_t * sam_account_name, wchar_t * user_principal_name)
```
please add a brief summary on what this routine does.

NoMoreFood replied:
Done

NoMoreFood committed:
Use Implicit User Principal Name If Explicit Does Not Exist - Revisions 1

- Moved lookup_principal_name() fallback behavior out of the function and into the caller.
- Added additional comments.
